src/main/java/com/google/sps/data/FirestoreDB.java

```diff
@@ -0,0 +1,259 @@
+    return getRecipeMetadataQuery(recipesQuery, sortingMethod);
+  }
+
+  public List<RecipeMetadata> getRecipesMatchingCreator(
+      String creatorId, SortingMethod sortingMethod) {
+    Query recipesQuery = DBUtils.recipes().whereEqualTo(Recipe.CREATOR_ID_KEY, creatorId);
+    return getRecipeMetadataQuery(recipesQuery, sortingMethod);
+  }
+  
+  public List<RecipeMetadata> getRecipesSavedBy(String userId, SortingMethod sortingMethod) {
+    List<String> saved_Ids = savedRecipeIds(userId);
+    return getRecipesMatchingIDs(saved_Ids, sortingMethod);
+  }
+
+  public List<RecipeMetadata> getRecipesMatchingIDs(List<String> Ids, SortingMethod sortingMethod) {
+    Query recipesQuery = DBUtils.recipes().whereIn(Recipe.ID_KEY, Ids);
+    return getRecipeMetadataQuery(recipesQuery, sortingMethod);
+  }
+  
+  public List<RecipeMetadata> getRecipeMetadataQuery(
+      Query recipesQuery, SortingMethod sortingMethod) {
+    switch (sortingMethod) {
+      case TOP:
+        recipesQuery = recipesQuery.orderBy(Recipe.VOTES_KEY, Query.Direction.DESCENDING);
+        break;
+      case NEW:
+        recipesQuery = recipesQuery.orderBy(Recipe.TIMESTAMP_KEY, Query.Direction.DESCENDING);
+        break;
+    }
+
+    ArrayList<RecipeMetadata> recipeList = new ArrayList<>();
+    QuerySnapshot querySnapshot = DBUtils.blockOnFuture(recipesQuery.get());
+
+    if (querySnapshot == null) {
+      return null;
+    }
+
+    for (DocumentSnapshot document : querySnapshot.getDocuments()) {
+      RecipeMetadata recipe = document.toObject(RecipeMetadata.class);
+      recipeList.add(recipe);
+    }
+    return recipeList;
+  }
+
+  public Query recipesMatchingTags(Iterable<String> tagIds, Iterator<String> iter) {
+    if (iter.hasNext()) {
+      String nextTag = iter.next();
+      return recipesMatchingTags(tagIds, iter).whereEqualTo("tags." + nextTag, true);
+    }
+    return DBUtils.recipes();
+  }
+
+  public List<String> savedRecipeIds(String userId) {
+    List<String> allRecipeIds = DBUtils.allRecipeIds();
+    ArrayList<String> savedRecipeIds = new ArrayList<String>();
+
+    for (String recipeID : allRecipeIds) {
+      if (isSavedRecipe(userId, recipeID)) {
+        savedRecipeIds.add(recipeID);
+      }
+    }
+    return savedRecipeIds;
+  }
+  
+  public String addComment(Comment newComment, String recipeId) {
+    DocumentReference newCommentRef = DBUtils.comments(recipeId).document();
+    DBUtils.blockOnFuture(newCommentRef.set(newComment));
+    return newCommentRef.getId();
+  }
+
+  public void deleteComment(String Id, String recipeId) {
+    DBUtils.blockOnFuture(DBUtils.comments(recipeId).document(Id).delete());
+  }
+
+  public void deleteComments(String recipeId) {
+    List<QueryDocumentSnapshot> documents =
+        DBUtils.blockOnFuture(DBUtils.comments(recipeId).get()).getDocuments();
+    for (QueryDocumentSnapshot document : documents) {
+      document.getReference().delete();
+    }
```

Comment:
Again, this is unnecessarily querying the actual documents, which we don't need to do. Iterate through the document references using `CollectionReference.listDocuments()` and delete the document references directly.